java/runtime/src/main/java/org/ray/runtime/actor/NativeRayActor.java

```diff
@@ -33,75 +44,64 @@ public NativeRayActor(long nativeActorHandle) {
   public NativeRayActor() {
   }
 
-  public long getNativeActorHandle() {
-    return nativeActorHandle;
-  }
-
   @Override
   public ActorId getId() {
-    return ActorId.fromBytes(nativeGetActorId(nativeActorHandle));
-  }
-
-  @Override
-  public UniqueId getHandleId() {
-    return new UniqueId(nativeGetActorHandleId(nativeActorHandle));
+    return ActorId.fromBytes(actorId);
   }
 
   public Language getLanguage() {
-    return Language.forNumber(nativeGetLanguage(nativeActorHandle));
+    return Language.forNumber(nativeGetLanguage(nativeCoreWorkerPointer, actorId));
   }
 
   public boolean isDirectCallActor() {
-    return nativeIsDirectCallActor(nativeActorHandle);
+    return nativeIsDirectCallActor(nativeCoreWorkerPointer, actorId);
   }
 
   @Override
   public String getModuleName() {
     Preconditions.checkState(getLanguage() == Language.PYTHON);
-    return nativeGetActorCreationTaskFunctionDescriptor(nativeActorHandle).get(0);
+    return nativeGetActorCreationTaskFunctionDescriptor(
+      nativeCoreWorkerPointer, actorId).get(0);
   }
 
   @Override
   public String getClassName() {
     Preconditions.checkState(getLanguage() == Language.PYTHON);
-    return nativeGetActorCreationTaskFunctionDescriptor(nativeActorHandle).get(1);
-  }
-
-  public NativeRayActor fork() {
-    return new NativeRayActor(nativeFork(nativeActorHandle));
+    return nativeGetActorCreationTaskFunctionDescriptor(
+      nativeCoreWorkerPointer, actorId).get(1);
   }
 
   @Override
   public void writeExternal(ObjectOutput out) throws IOException {
-    out.writeObject(nativeSerialize(nativeActorHandle));
+    out.writeObject(nativeSerialize(nativeCoreWorkerPointer, actorId));
   }
 
   @Override
   public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
-    nativeActorHandle = nativeDeserialize((byte[]) in.readObject());
+    RayRuntime runtime = Ray.internal();
+    if (runtime instanceof RayMultiWorkerNativeRuntime) {
+      runtime = ((RayMultiWorkerNativeRuntime) runtime).getCurrentRuntime();
+    }
+    
+    Preconditions.checkState(runtime instanceof RayNativeRuntime);
+    nativeCoreWorkerPointer = ((RayNativeRuntime)runtime).getNativeCoreWorkerPointer();
+
+    actorId = nativeDeserialize(nativeCoreWorkerPointer, (byte[]) in.readObject());
   }
 
   @Override
   protected void finalize() {
-    nativeFree(nativeActorHandle);
+    // TODO(zhijunfu): do we need to free the ActorHandle in core worker?
```

Comment:
@stephanie-wang It seems that we don't do this in Python either? If I understand it correctly, we only clean up all handles at the end of each non-actor task. 
I think this would lead to memory leak in some cases. E.g., an actor with a method that takes an actor handle as parameter.

Reply:
Hmm yeah that sounds right. Can you open an issue so we can track it? I don't think any of us have time to get to it right now.

Reply:
https://github.com/ray-project/ray/issues/6370